packages/contracts/contracts/CyberMarketplace.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import '@openzeppelin/contracts/utils/Counters.sol';
+import '@openzeppelin/contracts/utils/Address.sol';
+//import '@openzeppelin/contracts/access/Ownable.sol';
+import '@openzeppelin/contracts/token/ERC1155/IERC1155.sol';
+import '@openzeppelin/contracts/utils/introspection/ERC165Checker.sol';
+import '@openzeppelin/contracts/token/ERC1155/utils/ERC1155Holder.sol';
+
+bytes4 constant NFT_SIG = 0xd9b67a26;
+
+contract CyberMarketplace is ERC1155Holder {
+  using Counters for Counters.Counter;
+
+  struct Listing {
+    uint256 id;
+    address nftContract;
+    address payable seller;
+    uint256 tokenId;
+    uint256 price;
+    uint256 endTime;
+  }
+
+  event Sell(
+    uint256 id,
+    address nftContract,
+    uint256 tokenId,
+    address seller,
+    uint256 price,
+    uint256 endTime
+  );
+
+  event Buy(
+    uint256 id,
+    address nftContract,
+    uint256 tokenId,
+    address seller,
+    address buyer,
+    uint256 price
+  );
+
+  Counters.Counter private listingsId;
+
+  mapping(uint256 => Listing) private listings;
+
+  function sell(
+    address nftContract,
+    uint256 tokenId,
+    uint256 price,
+    uint256 duration
+  ) public {
+    // pre
+    require(
+      ERC165Checker.supportsInterface(nftContract, NFT_SIG),
+      'Invalid NFT'
+    );
+    require(
+      IERC1155(nftContract).balanceOf(msg.sender, tokenId) > 0,
+      'Empty balance'
+    );
+    require(
+      IERC1155(nftContract).isApprovedForAll(msg.sender, address(this)),
+      'Not approved'
+    );
+
+    listingsId.increment();
+    uint256 id = listingsId.current();
+
+    listings[id] = Listing(
+      id,
+      nftContract,
+      payable(msg.sender),
+      tokenId,
+      price,
+      block.timestamp + duration
+    );
+
+    IERC1155(nftContract).safeTransferFrom(
+      msg.sender,
+      address(this),
+      tokenId,
+      1,
+      '0x0'
+    );
+
+    emit Sell(
+      id,
+      nftContract,
+      tokenId,
+      msg.sender,
+      price,
+      block.timestamp + duration
+    );
+  }
+
+  function buy(uint256 listingId) public payable {
+    Listing memory listing = listings[listingId];
+    // pre
+    require(listing.id > 0, 'Invalid listing id');
+    require(block.timestamp < listing.endTime, 'Listing expired');
+    require(msg.value == listing.price, 'Invalid price');
+
+    delete listings[listingId];
```

Comment:
why delete the listing? I don't know if it's good or not, I want to know what's you motivation to perform that

Reply:
My thought was It's gonna take EVM storage unnecessarily (is there a limit to how much a contract can use?).  It makes sense to not delete it if we wanna keep this info for future use. But per the current spec I didn't see the use of it. 

What's the best practice in Solidity here, should we keep data in mapping even if it's obsolete?

Reply:
1/ No there is a contract (compiled bytecode of source) size limit, but data storage has no limit.
2/ I don't see why we will need to use an ended sells, maybe to show up history on some page, but even for the history we can use indexed service, then we don't need to keep it. Usually, data stored on-chain like stay there forever, especially since there is no discount to remove data, to delete will not offer the gas discount for the buyer, so we can keep it.
Idk if we will need it in the future, that is a good question yes.

So let's keep like this for now and kick off this question later on